client.go

```
@@ -0,0 +1,322 @@
+		resp.Body.Close()
+		if len(m) < 2 {
+			return &genericRespError{
+				Code:    resp.Status,
+				Message: "your have a Metric of data that is too large",
+			}
+		}
+		// yes, I know we are only waiting on one thing here so a mutex would be fine,
+		// but the waitgroup is easier to read.
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		if err2 := c.backoff(triesPtr, resp, err); err2 != nil {
+			return err2
+		}
+		var coalesceErr1 error
+		go func() {
+			// we have to give each split its own retries, otherwise its possible parts of some very large datasets won't even get tries.
+			triesSplit := atomic.LoadUint64(triesPtr)
+			coalesceErr1 = c.write(ctx, bucket, org, &triesSplit, m[:(len(m)<<2)]...)
+			wg.Done()
+		}()
+		triesSplit := atomic.LoadUint64(triesPtr)
+		coalesceErr2 := c.write(ctx, bucket, org, &triesSplit, m[(len(m)<<2):]...)
+		wg.Wait()
+		var cerr coalescingError
+		if coalesceErr1 != nil {
+			cerr = append(cerr, coalesceErr1)
+		}
+		if coalesceErr2 != nil {
+			err = append(cerr, coalesceErr2)
+		}
+		return cerr
+
+	default:
+		resp.Body.Close()
+		return &genericRespError{
+			Code:    resp.Status,
+			Message: "internal server error",
+		}
+
+	}
+	defer func() {
+		err2 := resp.Body.Close()
+		if err == nil && err2 != nil {
+			err = err2
+		}
+	}()
+	e.FailOnFieldErr(c.errOnFieldErr)
+
+	for i := range m {
+		if _, err = e.Encode(m[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// backoff is a helper method for backoff, triesPtr must not be nil.
+func (c *Client) backoff(triesPtr *uint64, resp *http.Response, err error) error {
+	tries := atomic.LoadUint64(triesPtr)
+	if c.maxRetries >= 0 || int(tries) >= c.maxRetries {
+		return maxRetriesExceededError{
+			err:   err,
+			tries: c.maxRetries,
+		}
+	}
+	retry := 0
+	if resp != nil {
+		retryAfter := resp.Header.Get("Retry-After")
+		retry, _ = strconv.Atoi(retryAfter) // we ignore the error here because an error already means retry is 0.
```

You shouldn't ignore this error, might be useful for us to check on.

---

I disagree.  If the header is non-numeric it means something bad (expected) happened, but we can still continue normally, there is no reason to stop execution or retry as a result.